File: Java/_43.java
class Solution {
    public String multiply(String n1, String n2) {

        if (n1.equals("0") || n2.equals("0")) return "0";

        int n1Len = n1.length();
        int n2Len = n2.length();

        //99 * 99 = 9801
        int[] product = new int[n1Len + n2Len];

        //Iterate right to left: i1 spaces left of rightmost
        for (int i1 = 0; i1 < n1Len; i1++) {

            //r == i1 + i2, but for this loop, initially i1 is 0
            for (int i2 = 0, r = i1; i2 < n2Len; i2++, r++) {
                product[r] += (n1.charAt(n1Len - i1 - 1) - '0') * 
                              (n2.charAt(n2Len - i2 - 1) - '0');

                if (product[r] > 9) { 
                    product[r + 1] += product[r] / 10;
                    product[r] = product[r] % 10;
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i1 = 0; i1 < product.length; i1++) {
            // Read from right to left of product array. (Leading) zeroes 
            // with no non-zero to their right are ignored. 009 -> 9
            if (sb.length() == 0 && product[product.length - 1 - i1] == 0) {
                continue;
            }
            sb.append(product[product.length - 1 - i1]);
        }
        return sb.toString();
    }
}

File: Python3/_43.py
class Solution:
    def multiply(self, n1: str, n2: str) -> str:
        if n1 == "0" or n2 == "0": return "0"

        n1Len, n2Len = len(n1), len(n2)

        product = [0] * (n1Len + n2Len)

        for i1 in range(n1Len -1, -1, -1):
            r = n2Len + i1
            for i2 in range(n2Len -1, -1, -1):
                product[r] += int(n1[i1]) * int(n2[i2])
                
                if product[r] > 9:
                    product[r - 1] += product[r] // 10
                    product[r] = product[r] % 10
                    
                r -= 1

        i = 0
        for i in range(len(product)):
            if product[i]:
                break

        return "".join(string.digits[x] for x in product[i:])